carma_wm/src/WMListenerWorker.cpp

```diff
@@ -93,6 +95,44 @@ void WMListenerWorker::mapCallback(const autoware_lanelet2_msgs::MapBinConstPtr&
   }
 }
 
+void WMListenerWorker::incomingSpatCallback(const cav_msgs::SPAT& spat_msg)
+{
+  if (spat_msg.intersection_state_list.empty())
+  {
+    ROS_WARN_STREAM("No intersection_state_list in the newly received SPAT msg. Returning...");
+    return;
+  }
+  for (auto curr_intersection : spat_msg.intersection_state_list)
+  {
+    for (auto current_movement_state : curr_intersection.movement_list) 
+    {
+      lanelet::Id curr_light_id = world_model_->getTrafficLightId(curr_intersection.id.id, current_movement_state.signal_group);
+      if (curr_light_id == lanelet::InvalId)
+      {
+        ROS_WARN_STREAM("Received a SPAT message for traffic light that is not in the map with intersection_id: " << curr_intersection.id.id << 
+                           ", and signal_group_id: " << current_movement_state.signal_group);
+        continue;
+      }
+      auto general_regem = world_model_->getMutableMap()->regulatoryElementLayer.get(curr_light_id);
+      
+      // get ptr to the regem in the map
+      lanelet::CarmaTrafficLightPtr curr_light = world_model_->getMutableMap()->laneletLayer.findUsages(general_regem)[0].regulatoryElementsAs<lanelet::CarmaTrafficLight>()[0];
+
+      // check if we have processed this already or not
+      if (curr_light->revision_ == curr_intersection.revision)
```

Comment:
I don't think the revision field works this way. I believe it only changes to the intersection structure not a new spat message. Per what we had discussed before you might need to see multiple Spat messages before the full phase is described.

Reply:
@aswath1 This is the disconnect. 
You are treating MovementEvent's as if each MovementEvent corresponds to one phase state. However, movements refer to something like what I have in the picture below
![image](https://user-images.githubusercontent.com/10292079/126706532-cf7b9934-3d5c-4d6e-90d9-5c31639e218f.png)

Each movement is one possible direction of travel from a lane through the intersection. So the SPaT message is actually only communicating the end time of the current state. You need to therefore listen to the message for the duration of a full cycle in order to populate the total set of phase transitions.